File: 716-max-stack/716-max-stack.js

var MaxStack = function() {
    this.stack = []
    this.maxstack = []
};

/** 
 * @param {number} x
 * @return {void}
 */
MaxStack.prototype.push = function(x) {
    let max = this.maxstack.length > 0 ? this.maxstack[this.maxstack.length-1] : x
    this.stack.push(x)
    this.maxstack.push(max > x ? max : x)
 };

/**
 * @return {number}
 */
MaxStack.prototype.pop = function() {
    this.maxstack.pop()
    return this.stack.pop()
};

/**
 * @return {number}
 */
MaxStack.prototype.top = function() {
    return this.stack[this.stack.length-1]
};

/**
 * @return {number}
 */
MaxStack.prototype.peekMax = function() {
    return this.maxstack[this.maxstack.length-1]
};

/**
 * @return {number}
 */
MaxStack.prototype.popMax = function() {
    let max = this.peekMax()
    let temp = []
    while(this.top() != max) {
        temp.push(this.pop())
    }
    this.pop()
    while(temp.length > 0) {
        this.push(temp.pop())
    }
    return max 
};

/** 
 * Your MaxStack object will be instantiated and called as such:
 * var obj = new MaxStack()
 * obj.push(x)
 * var param_2 = obj.pop()
 * var param_3 = obj.top()
 * var param_4 = obj.peekMax()
 * var param_5 = obj.popMax()
 */

File: 716-max-stack/716-max-stack.java
class MaxStack {
    
    private TreeSet<int[]> stack;
    private TreeSet<int[]> values;
    private int id = 0;

    public MaxStack() {
        stack = new TreeSet<int[]>((o1, o2) -> {
            return o1[1] - o2[1];
        });
        
        values = new TreeSet<int[]>((o1,o2) -> {
            if(o1[0] == o2[0]) {
                return o1[1] - o2[1];
            }
            return o1[0] - o2[0];
        });
    }
    
    public void push(int x) {
        id++;
        stack.add(new int[]{x, id});
        values.add(new int[] {x, id});
    }
    
    public int pop() {
        int[] pair = stack.pollLast();
        values.remove(pair);
        return pair[0];
    }
    
    public int top() {
        //System.out.println(stack);
        int[] pair = stack.last();
        //System.out.println(pair);
        return pair[0];
    }
    
    public int peekMax() {
        int[] pair = values.last();
        return pair[0];
    }
    
    public int popMax() {
        int[] pair = values.pollLast();
        stack.remove(pair);
        return pair[0];
    }
}

/**
 * Your MaxStack object will be instantiated and called as such:
 * MaxStack obj = new MaxStack();
 * obj.push(x);
 * int param_2 = obj.pop();
 * int param_3 = obj.top();
 * int param_4 = obj.peekMax();
 * int param_5 = obj.popMax();
 */